dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +232,412 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
-                
-                THROW_IF_FAILED(::CoRegisterClassObject(
-                    taskClsid,
-                    winrt::make<PushNotificationBackgroundTaskFactory>().get(),
-                    CLSCTX_LOCAL_SERVER,
-                    REGCLS_MULTIPLEUSE,
-                    &s_comActivatorRegistration));
-            }
+                winrt::guid registeredClsid{ GUID_NULL };
+                {
+                    auto lock{ m_lock.lock_shared() };
+                    registeredClsid = m_registeredClsid;
+                }
+                // registeredClsid is set to GUID_NULL for unpackaged applications
+                THROW_IF_FAILED(notificationsLongRunningPlatform->RegisterLongRunningActivatorWithClsid(processName.c_str(), registeredClsid));
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    m_singletonLongRunningSinkRegistration = true;
+                }
 
-            if (builder)
-            {
-                registeredTaskFromBuilder = builder.Register();
-            }
+                auto scopeExitToCleanLongRunningSink{ wil::scope_exit([&]()
+                {
+                    THROW_IF_FAILED(notificationsLongRunningPlatform->UnregisterLongRunningActivator(processName.c_str()));
+                    m_singletonLongRunningSinkRegistration = false;
+                }
+                ) };
 
-            scopeExitToCleanRegistrations.release();
-            auto lock{ s_activatorInfoLock.lock_exclusive() };
-            s_pushTriggerRegistration = registeredTaskFromBuilder;
-            PushNotificationTelemetry::ActivatorRegisteredByApi(S_OK, details.Activators());
-        }
+                // Register a COM object for the PushNotificationLongRunningProcess to CoCreate in background activation scenarios
+                if (AppModel::Identity::IsPackagedProcess())
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_IF_FAILED(::CoRegisterClassObject(
+                        registeredClsid,
+                        winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                        CLSCTX_LOCAL_SERVER,
+                        REGCLS_MULTIPLEUSE,
+                        &m_comActivatorRegistration));
+                }
 
-        catch(...)
+                scopeExitToCleanLongRunningSink.release();
+                scopeExitToCleanForegroundSink.release();
+                scopeExitFullTrustRegistration.release();
+                scopeExitToCleanRegistrations.release();
+            }
+        }
+        catch (...)
         {
-            PushNotificationTelemetry::ActivatorRegisteredByApi(wil::ResultFromCaughtException(),
-                details == nullptr ? PushNotificationRegistrationActivators::Undefined : details.Activators());
-            throw;
+            hr = wil::ResultFromCaughtException();
+            PushNotificationTelemetry::ActivatorRegisteredByApi(hr);
```

Comment:
you are logging twice. Here and below

Reply:
instead log once before you throw

Reply:
If it fails, it will log and throw outside the try/catch. Otherwise it will log success.